[DSM] DDP-6584 Downloading participant files

pegahtah opened the pull request:
I have added a new component `file-download` and a new tab in participant page to view the uploaded files. 
Also new columns for participant list has been added to enable searching on the fields
User must have permission and the participant must have uploaded files for the tab to appear.
<img width="1680" alt="Screen Shot 2022-07-26 at 11 50 32 AM" src="https://user-images.githubusercontent.com/47120870/181261299-62505ce8-99c9-4c7a-90c5-71c5e8a60f95.png">


<img width="265" alt="Screen Shot 2022-07-25 at 2 01 41 PM" src="https://user-images.githubusercontent.com/47120870/180857474-efffd7e6-9d39-42ac-aabe-4f71cd09b9e5.png">

Commit:
feat: DDP-6584 transferred HTML tab from old branch

Commit:
feat: DDP-6584 transfering old code to develop

Commit:
feat: DDP-6584 transfering old code to develop

pegahtah committed:
DDP-6584 some changes

pegahtah committed:
DDP-6584
adding file-download component
adding filter columns needed for filtering in participant-list in `column-model.ts` and `filter-columns.model.ts`
in `participant-list.component.html` the new display for columns are added
adding roles to give access to those who only have permissions

pegahtah committed:
DDP-6584
removing redundant code in participant-page.component.ts

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-6584_Download_files_new

pegahtah committed:
DDP-6584 linting fixes

pegahtah committed:
DDP-6584 removing unused import

alextanlan commented on ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html:
```diff
@@ -0,0 +1,38 @@
+<ng-container>
+  <ng-container *ngIf="downloading">
+    <div align="center">
```
`align` attribute is obsolete (see [link](https://developer.mozilla.org/en-US/docs/web/html/element/div#attr-align:~:text=Note%3A%20The%20align%20attribute%20is%20obsolete%3B%20do%20not%20use%20it%20anymore.%20Instead%2C%20you%20should%20use%20CSS%20properties%20or%20techniques%20such%20as%20CSS%20Grid%20or%20CSS%20Flexbox%20to%20align%20and%20position%20%3Cdiv%3E%20elements%20on%20the%20page.)).
Use css instead (text-align: center) .

alextanlan commented on ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html:
```diff
@@ -0,0 +1,38 @@
+<ng-container>
```
Wondering why there are so many `ng-container`-s here? 
Yu can get rid of all of them, like:
![image](https://user-images.githubusercontent.com/7396837/180981212-fa1f7d38-8215-4578-8c48-fe90d6d318e5.png)

alextanlan commented on ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html:
```diff
@@ -0,0 +1,38 @@
+        <td>
+          <button (click)="downloadParticipantFile(file)" [disabled]="file.scanResult !== 'CLEAN'"
+                  color="primary" mat-raised-button
+                  type="button">Download File
```
I believe `type=button` can be skipped.

alextanlan commented on ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts:
```diff
@@ -0,0 +1,66 @@
+  }
+
+
+  getNiceDateFormat( uploadedAt: any ): string {
```
`uploadedAt` should be a string according to `ddp-dsm-ui\src\app\participant-list\models\file.model.ts`.
Is not it?

alextanlan commented on ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html:
```diff
@@ -0,0 +1,38 @@
+        </th>
+      </tr>
+      </thead>
+      <tr *ngFor="let file of participant.data.files">
```
What if `participant` or `participant.data` are null/undefined ?

alextanlan commented on ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts:
```diff
@@ -0,0 +1,66 @@
+
+  }
+
+  hasRole(): RoleService {
```
the method `hasRole` is nowhere used (as well as `RoleService` itself).

alextanlan commented on ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html:
```diff
@@ -0,0 +1,38 @@
+        <td>{{getNiceDateFormat(file.uploadedAt)}}</td>
+        <td>{{file.scanResult}}</td>
+        <td>
+          <button (click)="downloadParticipantFile(file)" [disabled]="file.scanResult !== 'CLEAN'"
```
It makes sense to keep 'CLEAN' string as a constant.

alextanlan commented on ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts:
```diff
@@ -0,0 +1,66 @@
+      return;
+    }
+    const realm = localStorage.getItem( ComponentService.MENU_SELECTED_REALM );
+    this.dsmService.downloadParticipantFile( file.fileName, file.bucket, file.blobName, realm, file.mimeType ).subscribe(
```
Can you please add `unsubscribe` for the subscription?

pegahtah replied:
Thanks! I wasn't aware we have to unsubscribe as well

alextanlan replied:
https://blog.bitsrc.io/6-ways-to-unsubscribe-from-observables-in-angular-ab912819a78f

alextanlan commented on ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts:
```diff
@@ -0,0 +1,66 @@
+
+  saveParticipantFile( data: any, type: string, fileName: string ): void {
+    const blob = new Blob( [ data ], {type: type} );
+    const url = window.URL.createObjectURL( blob );
```
Just wondering what are lines 54-55 for?

alextanlan commented on ddp-workspace/projects/ddp-dsm-ui/src/app/participant-list/models/file.model.ts:
```diff
@@ -0,0 +1,18 @@
+    this.mimeType = mimeType;
+    this.scanResult = scanResult;
+    this.fileSize = fileSize;
+    this.blobName = blobName;
```
Lines 5-11 can be easily removed.

alextanlan commented on ddp-workspace/projects/ddp-dsm-ui/src/app/participant-list/models/data.model.ts:
```diff
@@ -21,6 +23,7 @@ export class Data {
     this.address = address;
     this.invitations = invitations;
     this.computed = computed;
+    this.files = files;
```
Lines 16-26 are not needed, can be deleted. 
Typescript does that job under the hood.

alextanlan commented on ddp-workspace/projects/ddp-dsm-ui/src/app/participant-list/participant-list.component.html:
```diff
@@ -541,6 +541,18 @@ <h3 class="Color--primary Line--Break">{{additionalMessage}}</h3>
+                    <ul class="NO-PADDING-START">
+                      <li class="LIST-SEPARATOR" *ngFor="let file of pt.data.files">
+                        <p class="BIT-OF-MARGIN">
+                          <ng-container>
```
I can not see any benefits for using ng-container here.

pegahtah replied:
As it seems, I unknowingly tend to over use them. Thanks for catching that!

alextanlan replied:
Often it is used when you have more than 1 structural directive on an element, e.g. ***ngIf** and ***ngFor**
see [examples](https://www.digitalocean.com/community/tutorials/angular-ng-container-element )

alextanlan commented on ddp-workspace/projects/ddp-dsm-ui/src/app/services/dsm.service.ts:
```diff
@@ -611,6 +611,17 @@ export class DSMService {
+    map.push( {name: 'fileName', value: fileName} );
+    map.push( {name: 'bucket', value: bucketName} );
+    map.push( {name: 'blob', value: blob} );
+    map.push( {name: 'responseType', value: 'blob'} );
```
**_#optional_**
The line is unnecessary. 
That parameter is already in `buildQueryBlobHeader` method.

alextanlan commented:
Can you please fix linter warnings?
![image](https://user-images.githubusercontent.com/7396837/181004289-c87c2e79-4adf-461d-a231-747ea58073e2.png)

pegahtah committed:
DDP-6584
PR comments: removing unused ng-container tags and code, adding unsubscribe to observable methods, check for npe and undefined

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-6584_Download_files_new

pegahtah committed:
DDP-6584
PR comments: lint fix

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-6584_Download_files_new

alextanlan committed:
Merge branch 'develop' into DDP-6584_Download_files_new
